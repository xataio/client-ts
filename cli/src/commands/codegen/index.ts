import { Flags } from '@oclif/core';
import { generate, isValidJavascriptTarget, javascriptTargets } from '@xata.io/codegen';
import chalk from 'chalk';
import { mkdir, writeFile } from 'fs/promises';
import path, { dirname, extname, relative } from 'path';
import { BaseCommand, ProjectConfig } from '../../base.js';

export const languages: Record<string, 'javascript' | 'typescript'> = {
  '.js': 'javascript',
  '.mjs': 'javascript',
  '.cjs': 'javascript',
  '.ts': 'typescript'
};

export const unsupportedExtensionError = (ext: string) => {
  return `Cannot generate code for a file with extension ${ext}. Please use one of the following extensions: ${Object.keys(
    languages
  ).join(', ')}`;
};

export default class Codegen extends BaseCommand {
  static description = 'Generate code from the current database schema';

  static examples = [];

  static flags = {
    ...this.commonFlags,
    ...this.databaseURLFlag,
    branch: this.branchFlag,
    output: Flags.string({
      char: 'o',
      description: 'Output file. Overwrites your project configuration setting'
    }),
    declarations: Flags.boolean({
      description:
        'Whether or not the declarations file should be generated. Overwrites your project configuration setting'
    }),
    'inject-branch': Flags.boolean({
      description:
        'Inject the branch name into the generated code. Useful if you have a build step and the branch name is not available at runtime'
    }),
    'javascript-output-target': Flags.string({
      description: 'The output target for the generated javascript code.'
    }),
    'worker-id': Flags.string({
      description: 'Xata worker deployment id'
    })
  };

  static args = [];

  async run(): Promise<void> {
    const { flags } = await this.parse(Codegen);
    const output = flags.output || this.projectConfig?.codegen?.output;
    const moduleType = this.projectConfig?.codegen?.moduleType;
    const javascriptTarget = flags['javascript-output-target'] || this.projectConfig?.codegen?.javascriptTarget;
    const workersBuildId = flags['worker-id'] || this.projectConfig?.codegen?.workersBuildId;

    if (!output) {
      return this.error(
        `Please, specify an output file as a flag or in your project configuration file first with ${chalk.bold(
          'xata config set codegen.output <path>'
        )}`
      );
    }

    if (javascriptTarget !== undefined && !isValidJavascriptTarget(javascriptTarget)) {
      return this.error(
        `Invalid javascript output target. Please use one of the following values: ${Object.keys(
          javascriptTargets
        ).join(', ')}`
      );
    }

    const ext = extname(output);
    const dir = dirname(output);
    const language = languages[ext];
    if (!language) {
      return this.error(unsupportedExtensionError(ext));
    }

    const xata = await this.getXataClient();
    const { workspace, region, database, branch, databaseURL } = await this.getParsedDatabaseURLWithBranch(
      flags.db,
      flags.branch
    );
    const branchDetails = await xata.api.branches.getBranchDetails({ workspace, region, database, branch });
    const { schema } = branchDetails;

    const codegenBranch = flags['inject-branch'] ? branch : undefined;
    const result = await generate({
      schema,
      databaseURL,
      language,
      moduleType,
      javascriptTarget,
      branch: codegenBranch,
      workspace,
      workersBuildId
    });

    const { typescript, javascript, types } = result;
    const code = language === 'typescript' ? typescript : javascript;

    await mkdir(dir, { recursive: true });
    await writeFile(output, code);
    if (types && (flags.declarations || this.projectConfig?.codegen?.declarations)) {
      await writeFile(path.join(dir, 'types.d.ts'), types);
    }

    this.success(`Your XataClient is generated at ./${relative(process.cwd(), output)}`);
  }

  static async runIfConfigured(projectConfig?: ProjectConfig) {
    if (projectConfig?.codegen?.output) return Codegen.run([]);
  }
}
