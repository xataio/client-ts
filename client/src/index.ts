import { errors } from './util/errors';

export interface XataRecord {
  id: string;
  xata: {
    version: number;
  };
  read(): Promise<this>;
  update(data: Selectable<this>): Promise<this>;
  delete(): Promise<void>;
}

export type Queries<T> = {
  [key in keyof T as T[key] extends Query<infer A, infer B> ? key : never]: T[key];
};

export type OmitQueries<T> = {
  [key in keyof T as T[key] extends Query<infer A, infer B> ? never : key]: T[key];
};

export type OmitLinks<T> = {
  [key in keyof T as T[key] extends XataRecord ? never : key]: T[key];
};

export type OmitMethods<T> = {
  // eslint-disable-next-line @typescript-eslint/ban-types
  [key in keyof T as T[key] extends Function ? never : key]: T[key];
};

export type Selectable<T> = Omit<OmitQueries<OmitMethods<T>>, 'id' | 'xata'>;

export type Select<T, K extends keyof T> = Pick<T, K> & Queries<T> & XataRecord;

export type Include<T> = {
  [key in keyof T as T[key] extends XataRecord ? key : never]?: boolean | Array<keyof Selectable<T[key]>>;
};

type SortDirection = 'asc' | 'desc';

type Operator =
  | '$gt'
  | '$lt'
  | '$ge'
  | '$le'
  | '$exists'
  | '$notExists'
  | '$endsWith'
  | '$startsWith'
  | '$pattern'
  | '$is'
  | '$isNot'
  | '$contains'
  | '$includes'
  | '$includesSubstring'
  | '$includesPattern'
  | '$includesAll';

// TODO: restrict constraints depending on type?
// E.g. startsWith cannot be used with numbers
type Constraint<T> = { [key in Operator]?: T };

type DeepConstraint<T> = T extends Record<string, any>
  ? {
      [key in keyof T]?: T[key] | DeepConstraint<T[key]>;
    }
  : Constraint<T>;

type ComparableType = number | Date;

export const gt = <T extends ComparableType>(value: T): Constraint<T> => ({ $gt: value });
export const ge = <T extends ComparableType>(value: T): Constraint<T> => ({ $ge: value });
export const gte = <T extends ComparableType>(value: T): Constraint<T> => ({ $ge: value });
export const lt = <T extends ComparableType>(value: T): Constraint<T> => ({ $lt: value });
export const lte = <T extends ComparableType>(value: T): Constraint<T> => ({ $le: value });
export const le = <T extends ComparableType>(value: T): Constraint<T> => ({ $le: value });
export const exists = (column: string): Constraint<string> => ({ $exists: column });
export const notExists = (column: string): Constraint<string> => ({ $notExists: column });
export const startsWith = (value: string): Constraint<string> => ({ $startsWith: value });
export const endsWith = (value: string): Constraint<string> => ({ $endsWith: value });
export const pattern = (value: string): Constraint<string> => ({ $pattern: value });
export const is = <T>(value: T): Constraint<T> => ({ $is: value });
export const isNot = <T>(value: T): Constraint<T> => ({ $isNot: value });
export const contains = <T>(value: T): Constraint<T> => ({ $contains: value });

// TODO: these can only be applied to columns of type "multiple"
export const includes = (value: string): Constraint<string> => ({ $includes: value });
export const includesSubstring = (value: string): Constraint<string> => ({ $includesSubstring: value });
export const includesPattern = (value: string): Constraint<string> => ({ $includesPattern: value });
export const includesAll = (value: string): Constraint<string> => ({ $includesAll: value });

type FilterConstraints<T> = {
  [key in keyof T]?: T[key] extends Record<string, any> ? FilterConstraints<T[key]> : T[key] | DeepConstraint<T[key]>;
};

type CursorNavigationOptions = { first?: string } | { last?: string } | { after?: string; before?: string };
type OffsetNavigationOptions = { size?: number; offset?: number };
type PaginationOptions = CursorNavigationOptions & OffsetNavigationOptions;

type BulkQueryOptions<T> = {
  page?: PaginationOptions;
  /** TODO: Not implemented yet
  filter?: FilterConstraints<T>;
  sort?:
    | {
        column: keyof T;
        direction?: SortDirection;
      }
    | keyof T;
**/
};

type QueryOrConstraint<T, R> = Query<T, R> | Constraint<T>;

type QueryMeta = { page: { cursor: string; more: boolean } };

interface BasePage<T, R> {
  query: Query<T, R>;
  meta: QueryMeta;
  records: R[];

  nextPage(size?: number, offset?: number): Promise<Page<T, R>>;
  previousPage(size?: number, offset?: number): Promise<Page<T, R>>;
  firstPage(size?: number, offset?: number): Promise<Page<T, R>>;
  lastPage(size?: number, offset?: number): Promise<Page<T, R>>;

  hasNextPage(): boolean;
}

class Page<T, R> implements BasePage<T, R> {
  readonly query: Query<T, R>;
  readonly meta: QueryMeta;
  readonly records: R[];

  constructor(query: Query<T, R>, meta: QueryMeta, records: R[] = []) {
    this.query = query;
    this.meta = meta;
    this.records = records;
  }

  async nextPage(size?: number, offset?: number): Promise<Page<T, R>> {
    return this.query.getPaginated({ page: { size, offset, after: this.meta.page.cursor } });
  }

  async previousPage(size?: number, offset?: number): Promise<Page<T, R>> {
    return this.query.getPaginated({ page: { size, offset, before: this.meta.page.cursor } });
  }

  async firstPage(size?: number, offset?: number): Promise<Page<T, R>> {
    return this.query.getPaginated({ page: { size, offset, first: this.meta.page.cursor } });
  }

  async lastPage(size?: number, offset?: number): Promise<Page<T, R>> {
    return this.query.getPaginated({ page: { size, offset, last: this.meta.page.cursor } });
  }

  // TODO: We need to add something on the backend if we want a hasPreviousPage
  hasNextPage(): boolean {
    return this.meta.page.more;
  }
}

export class Query<T, R = T> implements BasePage<T, R> {
  table: string;
  repository: Repository<T>;

  readonly $any?: QueryOrConstraint<T, R>[];
  readonly $all?: QueryOrConstraint<T, R>[];
  readonly $not?: QueryOrConstraint<T, R>[];
  readonly $none?: QueryOrConstraint<T, R>[];
  readonly $sort?: Record<string, SortDirection>;

  // Cursor pagination
  readonly query: Query<T, R> = this;
  readonly meta: QueryMeta = { page: { cursor: 'start', more: true } };
  readonly records: R[] = [];

  constructor(repository: Repository<T> | null, table: string, data: Partial<Query<T, R>>, parent?: Query<T, R>) {
    if (repository) {
      this.repository = repository;
    } else {
      this.repository = this as any;
    }
    this.table = table;

    // For some reason Object.assign(this, parent) didn't work in this case
    // so doing all this manually:
    this.$any = parent?.$any;
    this.$all = parent?.$all;
    this.$not = parent?.$not;
    this.$none = parent?.$none;
    this.$sort = parent?.$sort;

    Object.assign(this, data);
    // These bindings are used to support deconstructing
    // const { any, not, filter, sort } = xata.users.query()
    this.any = this.any.bind(this);
    this.all = this.all.bind(this);
    this.not = this.not.bind(this);
    this.filter = this.filter.bind(this);
    this.sort = this.sort.bind(this);
    this.none = this.none.bind(this);

    Object.defineProperty(this, 'table', { enumerable: false });
    Object.defineProperty(this, 'repository', { enumerable: false });
  }

  any(...queries: Query<T, R>[]): Query<T, R> {
    return new Query<T, R>(
      this.repository,
      this.table,
      {
        $any: (this.$any || []).concat(queries)
      },
      this
    );
  }

  all(...queries: Query<T, R>[]): Query<T, R> {
    return new Query<T, R>(
      this.repository,
      this.table,
      {
        $all: (this.$all || []).concat(queries)
      },
      this
    );
  }

  not(...queries: Query<T, R>[]): Query<T, R> {
    return new Query<T, R>(
      this.repository,
      this.table,
      {
        $not: (this.$not || []).concat(queries)
      },
      this
    );
  }

  none(...queries: Query<T, R>[]): Query<T, R> {
    return new Query<T, R>(
      this.repository,
      this.table,
      {
        $none: (this.$none || []).concat(queries)
      },
      this
    );
  }

  filter(constraints: FilterConstraints<T>): Query<T, R>;
  filter<F extends keyof T>(column: F, value: FilterConstraints<T[F]> | DeepConstraint<T[F]>): Query<T, R>;
  filter(a: any, b?: any): Query<T, R> {
    if (arguments.length === 1) {
      const constraints = a as FilterConstraints<T>;
      const queries: QueryOrConstraint<T, R>[] = [];
      for (const [column, constraint] of Object.entries(constraints)) {
        queries.push({ [column]: constraint });
      }
      return new Query<T, R>(
        this.repository,
        this.table,
        {
          $all: (this.$all || []).concat(queries)
        },
        this
      );
    } else {
      const column = a as keyof T;
      const value = b as Partial<T[keyof T]> | Constraint<T[keyof T]>;
      return new Query<T, R>(
        this.repository,
        this.table,
        {
          $all: (this.$all || []).concat({ [column]: value })
        },
        this
      );
    }
  }

  sort<F extends keyof T>(column: F, direction: SortDirection): Query<T, R> {
    const sort = { ...this.$sort, [column]: direction };
    const q = new Query<T, R>(
      this.repository,
      this.table,
      {
        $sort: sort
      },
      this
    );

    return q;
  }

  async getPaginated(options?: BulkQueryOptions<T>): Promise<Page<T, R>> {
    return this.repository._runQuery(this, options);
  }

  async *[Symbol.asyncIterator](): AsyncIterableIterator<R> {
    for await (const [record] of this.getIterator(1)) {
      yield record;
    }
  }

  async *getIterator(chunk: number, options: Omit<BulkQueryOptions<T>, 'page'> = {}): AsyncGenerator<R[]> {
    let offset = 0;
    let end = false;

    while (!end) {
      const { records, meta } = await this.getPaginated({ ...options, page: { size: chunk, offset } });
      yield records;

      offset += chunk;
      end = !meta.page.more;
    }
  }

  async getMany(options?: BulkQueryOptions<T>): Promise<R[]> {
    const { records } = await this.getPaginated(options);
    return records;
  }

  async getOne(options: Omit<BulkQueryOptions<T>, 'page'> = {}): Promise<R | null> {
    const records = await this.getMany({ ...options, page: { size: 1 } });
    return records[0] || null;
  }

  async deleteAll(): Promise<number> {
    // TODO: Return number of affected rows
    return 0;
  }

  include(columns: Include<T>) {
    // TODO
    return this;
  }

  async nextPage(size?: number, offset?: number): Promise<Page<T, R>> {
    return this.firstPage(size, offset);
  }

  async previousPage(size?: number, offset?: number): Promise<Page<T, R>> {
    return this.firstPage(size, offset);
  }

  async firstPage(size?: number, offset?: number): Promise<Page<T, R>> {
    return this.getPaginated({ page: { size, offset } });
  }

  async lastPage(size?: number, offset?: number): Promise<Page<T, R>> {
    return this.getPaginated({ page: { size, offset, before: 'end' } });
  }

  hasNextPage(): boolean {
    return this.meta.page.more;
  }
}

export abstract class Repository<T> extends Query<T, Selectable<T>> {
  select<K extends keyof Selectable<T>>(...columns: K[]) {
    return new Query<T, Select<T, K>>(this.repository, this.table, {});
  }

  abstract create(object: Selectable<T>): Promise<T>;

  abstract createMany(objects: Selectable<T>[]): Promise<T[]>;

  abstract read(id: string): Promise<T | null>;

  abstract update(id: string, object: Partial<T>): Promise<T>;

  abstract delete(id: string): void;

  // Used by the Query object internally
  abstract _runQuery<R>(query: Query<T, R>, options?: BulkQueryOptions<T>): Promise<Page<T, R>>;
}

export class RestRepository<T> extends Repository<T> {
  client: BaseClient<any>;
  fetch: any;

  constructor(client: BaseClient<any>, table: string) {
    super(null, table, {});
    this.client = client;

    const doWeHaveFetch = typeof fetch !== 'undefined';
    const isInjectedFetchProblematic = !this.client.options.fetch;

    if (doWeHaveFetch) {
      this.fetch = fetch;
    } else if (isInjectedFetchProblematic) {
      throw new Error(errors.falsyFetchImplementation);
    } else {
      this.fetch = this.client.options.fetch;
    }

    Object.defineProperty(this, 'client', { enumerable: false });
    Object.defineProperty(this, 'fetch', { enumerable: false });
    Object.defineProperty(this, 'hostname', { enumerable: false });
  }

  async request<T>(method: string, path: string, body?: unknown): Promise<T | undefined> {
    const { databaseURL, apiKey } = this.client.options;
    const branch = await this.client.getBranch();
    const fetchImpl = this.fetch;

    const resp: Response = await fetchImpl(`${databaseURL}:${branch}${path}`, {
      method,
      headers: {
        Accept: '*/*',
        'Content-Type': 'application/json',
        Authorization: `Bearer ${apiKey}`
      },
      body: JSON.stringify(body)
    });

    if (!resp.ok) {
      try {
        const json = await resp.json();
        const message = json.message;
        if (typeof message === 'string') {
          throw new XataError(message, resp.status);
        }
      } catch (err) {
        if (err instanceof XataError) throw err;
        // Ignore errors for other reasons.
        // For example if the response's body cannot be parsed as JSON
      }
      throw new XataError(resp.statusText, resp.status);
    }

    if (resp.status === 204) return undefined;
    return resp.json();
  }

  select<K extends keyof T>(...columns: K[]) {
    return new Query<T, Select<T, K>>(this.repository, this.table, {});
  }

  async create(object: T): Promise<T> {
    const record = transformObjectLinks(object);

    const response = await this.request<{
      id: string;
      xata: { version: number };
    }>('POST', `/tables/${this.table}/data`, record);
    if (!response) {
      throw new Error("The server didn't return any data for the query");
    }

    const finalObject = await this.read(response.id);
    if (!finalObject) {
      throw new Error('The server failed to save the record');
    }

    return finalObject;
  }

  async createMany(objects: T[]): Promise<T[]> {
    const records = objects.map((object) => transformObjectLinks(object));

    const response = await this.request<{
      recordIDs: string[];
    }>('POST', `/tables/${this.table}/bulk`, { records });
    if (!response) {
      throw new Error("The server didn't return any data for the query");
    }

    // TODO: Use filer.$any() to get all the records
    const finalObjects = await Promise.all(response.recordIDs.map((id) => this.read(id)));
    if (finalObjects.some((object) => !object)) {
      throw new Error('The server failed to save the record');
    }

    return finalObjects as T[];
  }

  async read(id: string): Promise<T | null> {
    try {
      const response = await this.request<
        T & { id: string; xata: { version: number; table?: string; warnings?: string[] } }
      >('GET', `/tables/${this.table}/data/${id}`);
      if (!response) return null;

      return this.client.initObject(this.table, response);
    } catch (err) {
      if ((err as XataError).status === 404) return null;
      throw err;
    }
  }

  async update(id: string, object: Partial<T>): Promise<T> {
    const response = await this.request<{
      id: string;
      xata: { version: number };
    }>('PUT', `/tables/${this.table}/data/${id}`, object);
    if (!response) {
      throw new Error("The server didn't return any data for the query");
    }

    // TODO: Review this, not sure we are properly initializing the object
    return this.client.initObject(this.table, response);
  }

  async delete(id: string) {
    await this.request('DELETE', `/tables/${this.table}/data/${id}`);
  }

  async _runQuery<R>(query: Query<T, R>, options?: BulkQueryOptions<T>): Promise<Page<T, R>> {
    const filter = {
      $any: query.$any,
      $all: query.$all,
      $not: query.$not,
      $none: query.$none
    };

    const body = {
      filter: Object.values(filter).some(Boolean) ? filter : undefined,
      sort: query.$sort,
      page: options?.page
    };

    const response = await this.request<{
      records: object[];
      meta: { page: { cursor: string; more: boolean } };
    }>('POST', `/tables/${this.table}/query`, body);
    if (!response) {
      throw new Error("The server didn't return any data for the query");
    }

    const { meta, records: objects } = response;
    const records = objects.map((record) => this.client.initObject<R>(this.table, record));

    return new Page(query, meta, records);
  }
}

interface RepositoryFactory {
  createRepository<T>(client: BaseClient<any>, table: string): Repository<T>;
}

export class RestRespositoryFactory implements RepositoryFactory {
  createRepository<T>(client: BaseClient<any>, table: string): Repository<T> {
    return new RestRepository<T>(client, table);
  }
}

type BranchStrategyValue = string | undefined | null;
type BranchStrategyBuilder = () => BranchStrategyValue | Promise<BranchStrategyValue>;
type BranchStrategy = BranchStrategyValue | BranchStrategyBuilder;
type BranchStrategyOption = NonNullable<BranchStrategy | BranchStrategy[]>;

export type XataClientOptions = {
  fetch?: unknown;
  databaseURL?: string;
  branch: BranchStrategyOption;
  apiKey: string;
  repositoryFactory?: RepositoryFactory;
};

export class BaseClient<D extends Record<string, Repository<any>>> {
  options: XataClientOptions;
  private links: Links;
  private branch: BranchStrategyValue;
  db!: D;

  constructor(options: XataClientOptions, links: Links) {
    if (!options.databaseURL || !options.apiKey || !options.branch) {
      throw new Error('Options databaseURL, apiKey and branch are required');
    }

    this.options = options;
    this.links = links;
  }

  public initObject<T>(table: string, object: object) {
    const o: Record<string, unknown> = {};
    Object.assign(o, object);

    const tableLinks = this.links[table] || [];
    for (const link of tableLinks) {
      const [field, linkTable] = link;
      const value = o[field];

      if (value && typeof value === 'object') {
        const { id } = value as any;
        if (Object.keys(value).find((col) => col === 'id')) {
          o[field] = this.initObject(linkTable, value);
        } else if (id) {
          o[field] = {
            id,
            get: () => {
              this.db[linkTable].read(id);
            }
          };
        }
      }
    }

    const db = this.db;
    o.read = function () {
      return db[table].read(o['id'] as string);
    };
    o.update = function (data: any) {
      return db[table].update(o['id'] as string, data);
    };
    o.delete = function () {
      return db[table].delete(o['id'] as string);
    };

    for (const prop of ['read', 'update', 'delete']) {
      Object.defineProperty(o, prop, { enumerable: false });
    }

    // TODO: links and rev links

    Object.freeze(o);
    return o as T;
  }

  public async getBranch(): Promise<string> {
    if (this.branch) return this.branch;

    const { branch: param } = this.options;
    const strategies = Array.isArray(param) ? [...param] : [param];

    const evaluateBranch = async (strategy: BranchStrategy) => {
      return isBranchStrategyBuilder(strategy) ? await strategy() : strategy;
    };

    for await (const strategy of strategies) {
      const branch = await evaluateBranch(strategy);
      if (branch) {
        this.branch = branch;
        return branch;
      }
    }

    throw new Error('Unable to resolve branch value');
  }
}

export class XataError extends Error {
  readonly status: number;

  constructor(message: string, status: number) {
    super(message);
    this.status = status;
  }
}

export type Links = Record<string, Array<string[]>>;

const isBranchStrategyBuilder = (strategy: BranchStrategy): strategy is BranchStrategyBuilder => {
  return typeof strategy === 'function';
};

// TODO: We can find a better implementation for links
const transformObjectLinks = (object: any) => {
  return Object.entries(object).reduce((acc, [key, value]) => {
    if (value && typeof value === 'object' && typeof (value as Record<string, unknown>).id === 'string') {
      return { ...acc, [key]: (value as XataRecord).id };
    }

    return { ...acc, [key]: value };
  }, {});
};
