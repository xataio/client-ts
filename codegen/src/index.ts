export { generate } from './codegen.js';
export type { GenerateOptions, Language } from './codegen.js';
export type { XataConfigSchema } from './config.js';
export { parseConfigFile } from './parseConfigFile.js';
export { parseSchemaFile } from './parseSchemaFile.js';
export type { Column, Table, XataDatabaseSchema } from './schema.js';
