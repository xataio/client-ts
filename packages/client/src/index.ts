export class XataError extends Error {
  readonly status: number;

  constructor(message: string, status: number) {
    super(message);
    this.status = status;
  }
}

export * from './api';
export * from './plugins';
export * from './client';
export * from './schema';
export * from './search';
export * from './serializer';
export * from './util/config';
export * from './util/apiKey';
export * from './workers';
export * from './version';
