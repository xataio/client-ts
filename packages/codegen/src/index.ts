export { generate, isValidJavascriptTarget, javascriptTargets } from './codegen';
export type { GenerateOptions, Language, ModuleType } from './codegen';
export { parseSchemaFile, columnSchema, tableSchema, revlinkSchema } from './schema';
export type { Column, Table, XataDatabaseSchema } from './schema';
