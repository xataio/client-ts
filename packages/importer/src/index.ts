export { guessColumnTypes } from './columns';
export { DEFAULT_CSV_DELIMITERS_TO_GUESS } from './parsers/csvParser';
export * from './plugin';
export type { CsvResults, ImportError, ImportBatchOptions, ParseMeta, ParseResults } from './types';
export { generateRandomData } from './random-data';
export { importColumnTypes } from './constants';
