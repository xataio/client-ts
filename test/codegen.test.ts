import { afterEach, describe, expect, it, vi } from 'vitest';
import { generate } from '../packages/codegen/src/codegen';
import { VERSION as CODEGEN_VERSION } from '../packages/codegen/src/version';

afterEach(() => {
  vi.clearAllMocks();
});

describe('generate', () => {
  it('should respect numbers in names', async () => {
    const output = await generate({
      schema: {
        tables: [
          {
            name: '1teams-case',
            primaryKey: ['xata_id'],
            columns: [
              { name: '2nameCase', type: 'string' },
              { name: '3Labels', type: 'multiple' }
            ]
          }
        ]
      },
      language: 'typescript',
      databaseURL: 'https://workspace-1234.xata.sh/db/dbname'
    });

    expect(stableVersion(output.typescript)).toMatchSnapshot();
    expect(stableVersion(output.javascript)).toMatchSnapshot();
    expect(stableVersion(output.types ?? '')).toMatchSnapshot();
  });

  it('should respect case naming', async () => {
    const output = await generate({
      schema: {
        tables: [
          {
            name: 'teams_Like',
            primaryKey: ['xata_id'],
            columns: [
              { name: 'name-test', type: 'string' },
              { name: 'labels_Test', type: 'multiple' },
              { name: 'ownerFoo', type: 'link', link: { table: 'users-foo' } }
            ]
          },
          {
            name: 'users-foo',
            primaryKey: ['xata_id'],
            columns: [
              { name: 'email-random', type: 'email' },
              { name: 'full_name', type: 'string' },
              { name: 'teamLike', type: 'link', link: { table: 'teams_Like' } }
            ]
          }
        ]
      },
      language: 'typescript',
      databaseURL: 'https://workspace-1234.xata.sh/db/dbname'
    });

    expect(stableVersion(output.typescript)).toMatchSnapshot();
  });

  it('should inject branch if passed', async () => {
    const output = await generate({
      schema: {
        tables: [
          {
            name: 'users',
            primaryKey: ['xata_id'],
            columns: [{ name: 'name', type: 'string' }]
          }
        ]
      },
      language: 'typescript',
      databaseURL: 'https://workspace-1234.xata.sh/db/dbname',
      branch: 'feature-branch'
    });

    expect(stableVersion(output.typescript)).toMatchSnapshot();
  });

  it('should generate CJS code', async () => {
    const output = await generate({
      schema: {
        tables: [
          {
            name: 'users',
            primaryKey: ['xata_id'],
            columns: [{ name: 'name', type: 'string' }]
          }
        ]
      },
      language: 'javascript',
      moduleType: 'cjs',
      databaseURL: 'https://workspace-1234.xata.sh/db/dbname',
      branch: 'feature-branch'
    });

    expect(stableVersion(output.javascript)).toMatchSnapshot();
  });

  it('should ignore CJS for TS code', async () => {
    const output = await generate({
      schema: {
        tables: [
          {
            name: 'users',
            primaryKey: ['xata_id'],
            columns: [{ name: 'name', type: 'string' }]
          }
        ]
      },
      language: 'typescript',
      moduleType: 'cjs',
      databaseURL: 'https://workspace-1234.xata.sh/db/dbname',
      branch: 'feature-branch'
    });

    expect(stableVersion(output.typescript)).toMatchSnapshot();
  });

  it('should generate Deno code', async () => {
    const output = await generate({
      schema: {
        tables: [
          {
            name: 'users',
            primaryKey: ['xata_id'],
            columns: [{ name: 'name', type: 'string' }]
          }
        ]
      },
      language: 'typescript',
      moduleType: 'deno',
      databaseURL: 'https://workspace-1234.xata.sh/db/dbname',
      branch: 'feature-branch'
    });

    expect(stableVersion(output.typescript)).toMatchSnapshot();
  });
});

// This method will replace the current version of the codegen with a placeholder so snapshots don't change
// if we just bump the codegen version
function stableVersion(code: string) {
  return code.replace(
    `Generated by Xata Codegen ${CODEGEN_VERSION}. Please do not edit`,
    'Generated by Xata Codegen CODEGEN_VERSION. Please do not edit'
  );
}
