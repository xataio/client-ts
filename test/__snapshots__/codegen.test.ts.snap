// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen should generate correct JavaScript 1`] = `
"/** @typedef { import('@xata.io/client').Repository } Repository */
import { BaseClient, RestRespositoryFactory } from \\"@xata.io/client\\";
/**
 * @typedef {Object} Author
 * @property {string} id
 * @property {Object} xata
 * @property {() => Promise<Author>} read
 * @property {() => Promise<Author>} update
 * @property {() => Promise<void>} delete
 * @property {string=} name
 * @property {string=} email
 * @property {string=} photoUrl
 * @property {string=} bio
 * @property {string=} title
 
 */
/**
 * @typedef {Object} Post
 * @property {string} id
 * @property {Object} xata
 * @property {() => Promise<Post>} read
 * @property {() => Promise<Post>} update
 * @property {() => Promise<void>} delete
 * @property {string=} title
 * @property {string=} summary
 * @property {string=} content
 * @property {string[]=} tags
 * @property {number=} likes
 * @property {AuthorRecord=} author
 * @property {string=} date
 * @property {string=} slug
 * @property {boolean=} published
 
 */
/**
 * @typedef {Object} Person
 * @property {string} id
 * @property {Object} xata
 * @property {() => Promise<Person>} read
 * @property {() => Promise<Person>} update
 * @property {() => Promise<void>} delete
 * @property {string=} slug
 * @property {string=} name
 * @property {string=} title
 * @property {string=} photoUrl
 
 */
const links = { authors: [], posts: [[\\"author\\", \\"authors\\"]], people: [] };
export class XataClient extends BaseClient {
  constructor(options) {
    super(
      {
        databaseURL: \\"https://my-workspace-gfdd12.xata.sh/db/hello-xata\\",
        ...options,
      },
      links
    );
    const factory = options.repositoryFactory || new RestRespositoryFactory();
    /** @type {{ \\"authors\\": Repository; \\"posts\\": Repository; \\"people\\": Repository }} */
    this.db = {
      authors: factory.createRepository(this, \\"authors\\"),
      posts: factory.createRepository(this, \\"posts\\"),
      people: factory.createRepository(this, \\"people\\"),
    };
  }
}
"
`;

exports[`codegen should generate correct TypeScript 1`] = `
"import {
  BaseClient,
  Repository,
  RestRespositoryFactory,
  XataClientOptions,
  XataRecord,
} from \\"@xata.io/client\\";

export interface Author {
  name?: string | null;
  email?: string | null;
  photoUrl?: string | null;
  bio?: string | null;
  title?: string | null;
}

export type AuthorRecord = Author & XataRecord;

export interface Post {
  title?: string | null;
  summary?: string | null;
  content?: string | null;
  tags?: string[] | null;
  likes?: number | null;
  author?: AuthorRecord | null;
  date?: string | null;
  slug?: string | null;
  published?: boolean | null;
}

export type PostRecord = Post & XataRecord;

export interface Person {
  slug?: string | null;
  name?: string | null;
  title?: string | null;
  photoUrl?: string | null;
}

export type PersonRecord = Person & XataRecord;

const links = { authors: [], posts: [[\\"author\\", \\"authors\\"]], people: [] };

export class XataClient extends BaseClient<{
  authors: Repository<Author>;
  posts: Repository<Post>;
  people: Repository<Person>;
}> {
  constructor(options: XataClientOptions) {
    super(
      {
        databaseURL: \\"https://my-workspace-gfdd12.xata.sh/db/hello-xata\\",
        ...options,
      },
      links
    );

    const factory = options.repositoryFactory || new RestRespositoryFactory();

    this.db = {
      authors: factory.createRepository(this, \\"authors\\"),
      posts: factory.createRepository(this, \\"posts\\"),
      people: factory.createRepository(this, \\"people\\"),
    };
  }
}
"
`;
